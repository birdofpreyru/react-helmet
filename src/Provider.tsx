import {
  type FunctionComponent,
  type ReactNode,
  createContext,
  useRef,
} from 'react';

import type {
  ContextValue,
  HelmetDataContext,
  HelmetProps,
  HelmetProviderHeap,
} from './types';

import { newServerState } from './server';
import { IS_DOM_ENVIRONMENT } from './constants';
import { calcAggregatedState } from './utils';
import { commitTagChanges } from './client';

export const Context = createContext<ContextValue | undefined>(undefined);

type ProviderProps = {
  children?: ReactNode;
  context?: HelmetDataContext;
};

const HelmetProvider: FunctionComponent<ProviderProps> = ({
  children,
  context,
}) => {
  const { current: heap } = useRef<HelmetProviderHeap>({
    firstRender: true,
    helmets: [],
    state: undefined,
  });

  const contextValueRef = useRef<ContextValue>(null);

  contextValueRef.current ??= {
    clientApply() {
      if (IS_DOM_ENVIRONMENT && !heap.state) {
        heap.state = calcAggregatedState(heap.helmets);
        if (heap.state.defer) {
          heap.nextAnimFrameId ??= requestAnimationFrame(() => {
            heap.state ??= calcAggregatedState(heap.helmets);
            commitTagChanges(heap.state, heap.firstRender);
            heap.firstRender = false;
            delete heap.nextAnimFrameId;
          });
        } else {
          if (heap.nextAnimFrameId !== undefined) {
            cancelAnimationFrame(heap.nextAnimFrameId);
            delete heap.nextAnimFrameId;
          }
          commitTagChanges(heap.state, heap.firstRender);
          heap.firstRender = false;
        }
      }
    },
    update(id: string, props: HelmetProps | undefined) {
      const idx = heap.helmets.findIndex((item) => item[0] === id);
      if (idx >= 0) {
        delete heap.state;
        if (props) heap.helmets[idx]![1] = props;
        else heap.helmets.splice(idx, 1);
      } else if (props) {
        delete heap.state;
        heap.helmets.push([id, props]);
      }
    },
  };

  if (context && (!context.helmet || context.helmet !== heap.serverState)) {
    heap.serverState ??= newServerState(heap);

    // eslint-disable-next-line no-param-reassign
    context.helmet = heap.serverState;
  }

  return <Context value={contextValueRef.current}>{children}</Context>;
};

export default HelmetProvider;
