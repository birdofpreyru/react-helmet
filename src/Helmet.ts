import {
  type FunctionComponent,
  type ReactElement,
  type ReactNode,
  Children,
  use,
  useEffect,
  useId,
} from 'react';

import { REACT_TAG_MAP, TAG_NAMES, VALID_TAG_NAMES } from './constants';

import { Context } from './Provider';

import type {
  BaseProps,
  BodyProps,
  HelmetChildProps,
  HelmetProps,
  HtmlProps,
  LinkProps,
  MetaProps,
  NoscriptProps,
  ScriptProps,
  StyleProps,
  TitleProps,
} from './types';
import { cloneProps, mergeProps, pushToPropArray } from './utils';

function assertChildType(
  childType: ReactElement['type'],
  nestedChildren: ReactNode,
): asserts childType is TAG_NAMES {
  if (typeof childType !== 'string') {
    throw Error(
      'You may be attempting to nest <Helmet> components within each other, which is not allowed. Refer to our API for more information.',
    );
  }

  if (!(VALID_TAG_NAMES as string[]).includes(childType)) {
    throw Error(
      `Only elements types ${VALID_TAG_NAMES.join(', ')} are allowed. Helmet does not support rendering <${childType}> elements. Refer to our API for more information.`,
    );
  }

  if (
    !nestedChildren
    || typeof nestedChildren === 'string'
    || Array.isArray(nestedChildren)
    // TODO: This piece of the check is wrong when parent is a fragment,
    // and thus children may not be an array of strings.
    // && nestedChildren.every((item) => typeof item === 'string')
  ) return;

  throw Error(
    `Helmet expects a string as a child of <${childType}>. Did you forget to wrap your children in braces? ( <${childType}>{\`\`}</${childType}> ) Refer to our API for more information.`,
  );
}

/**
 * Given a string key, it checks it against the legacy mapping between supported
 * HTML attribute names and their corresponding React prop names (for the names
 * that are different). If found in the mapping, it prints a warning to console
 * and returns the mapped prop name. Otherwise, it just returns the key as is,
 * assuming it is already a valid React prop name.
 */
function getPropName(key: string): keyof HelmetChildProps {
  const res = REACT_TAG_MAP[key];
  if (res) {
    // eslint-disable-next-line no-console
    console.warn(
      `"${key}" is not a valid JSX prop, replace it by "${res}"`,
    );
  }
  return (res ?? key) as keyof HelmetChildProps;
}

/**
 * Given children and props of a <Helmet> component, it reduces them to a single
 * props object.
 *
 * TODO: I guess, it should be further refactored, to make it cleaner...
 * though, it should perfectly work as is, so not a huge priority for now.
 */
function reduceChildrenAndProps(props: HelmetProps): Omit<HelmetProps, 'children'> {
  // NOTE: `props` are clonned, thus it is safe to push additional items to
  // array values of `res`, and to re-assign non-array values of `res`, without
  // the risk to mutate the original `props` object.
  const res: HelmetProps = cloneProps(props);

  // TODO: This is a temporary block, for compatibility with legacy library.
  for (const item of Object.values(props)) {
    if (Array.isArray(item)) {
      for (const it of item) {
        // TODO: This condition is actually needed to prevent some test failures,
        // I guess, something is messed up with related types?
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        if (it) {
          for (const key of Object.keys(it)) {
            const p = getPropName(key);
            if (p !== key) {
              it[p] = it[key as keyof HelmetChildProps] as unknown;
              delete it[key as keyof HelmetChildProps];
            }
          }
        }
      }
    } else if (item && typeof item === 'object') {
      const it = item as HelmetChildProps;
      for (const key of Object.keys(it)) {
        const p = getPropName(key);
        if (p !== key) {
          it[p] = it[key as keyof HelmetChildProps] as unknown;
          delete it[key as keyof HelmetChildProps];
        }
      }
    }
  }

  // eslint-disable-next-line complexity
  Children.forEach(props.children, (child) => {
    if (child === undefined || child === null) return;

    if (typeof child !== 'object' || !('props' in child)) {
      throw Error(
        `"${typeof child}" is not a valid <Helmet> descendant`,
      );
    }

    let nestedChildren: ReactNode;
    const childProps: HelmetChildProps = {};
    if (child.props) {
      for (const [key, value] of Object.entries(child.props)) {
        if (key === 'children') nestedChildren = value as ReactNode;
        else childProps[getPropName(key)] = value as unknown;
      }
    }

    let { type } = child;
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-conversion
    if (typeof type === 'symbol') type = (type as 'symbol').toString();
    assertChildType(type, nestedChildren);

    function assertStringChild(child2: ReactNode): asserts child2 is string {
      if (typeof child2 !== 'string') {
        // TODO: We want to throw, but the legacy code did not, so we won't for
        // now.
        // eslint-disable-next-line no-console
        console.error(`child of ${type as string} element should be a string`);

        /*
        throw Error(
          // NOTE: assertChildType() above guarantees that `type` is a string,
          // although it is not expressed in a way TypeScript can automatically
          // pick up.
        );
        */
      }
    }

    switch (type) {
      case TAG_NAMES.BASE:
        res.base = childProps as BaseProps;
        break;

      case TAG_NAMES.BODY:
        res.bodyAttributes = childProps as BodyProps;
        break;

      case TAG_NAMES.FRAGMENT:
        mergeProps(res, reduceChildrenAndProps({ children: nestedChildren }));
        break;

      case TAG_NAMES.HTML:
        res.htmlAttributes = childProps as HtmlProps;
        break;

      case TAG_NAMES.LINK:
      case TAG_NAMES.META:
        if (nestedChildren) {
          throw Error(
            `<${type} /> elements are self-closing and can not contain children. Refer to our API for more information.`,
          );
        }
        pushToPropArray(res, type, childProps as LinkProps | MetaProps);
        break;

      case TAG_NAMES.NOSCRIPT:
      case TAG_NAMES.SCRIPT:
        if (nestedChildren !== undefined) {
          assertStringChild(nestedChildren);
          (childProps as NoscriptProps | ScriptProps)
            .innerHTML = nestedChildren;
        }
        pushToPropArray(res, type, childProps);
        break;

      case TAG_NAMES.STYLE:
        assertStringChild(nestedChildren);
        (childProps as StyleProps).cssText = nestedChildren;
        pushToPropArray(res, type, childProps as StyleProps);
        break;

      case TAG_NAMES.TITLE:
        res.titleAttributes = childProps as TitleProps;

        if (typeof nestedChildren === 'string') res.title = nestedChildren;

        // When title contains {} expressions the children are an array of
        // strings, and other values.
        else if (Array.isArray(nestedChildren)) res.title = nestedChildren.join('');
        break;

      case TAG_NAMES.HEAD:
      default: {
        // TODO: Perhaps, we should remove HEAD entry from TAG_NAMES?
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const bad: TAG_NAMES.HEAD = type;
      }
    }
  });

  delete res.children;
  return res;
}

const Helmet: FunctionComponent<HelmetProps> = (props) => {
  const context = use(Context);

  if (!context) {
    throw Error(
      '<Helmet> component must be within a <HelmetProvider> children tree',
    );
  }

  const id = useId();

  // TODO: Agh... we need it here to ensure that it works server-side,
  // and we need the same in the useEffect() below to ensure it works
  // client-side in strict mode (and, thus completely correctly from React's
  // pure component / side effect logic). It clearly should be optimized,
  // but let's care about it later.
  context.update(id, reduceChildrenAndProps(props));

  // TODO: I guess, these two useEffect() can be merged together, which should
  // also allow to simplify and optimize the client-side management of attrs and
  // elements managed by these. Though, keeping them separate is an easier way
  // for now to ensure backward compatibility.
  useEffect(() => {
    context.update(id, reduceChildrenAndProps(props));
    context.clientApply();
  });

  useEffect(() => () => {
    context.update(id, undefined);
  }, [context, id]);

  return null;
};

export default Helmet;
