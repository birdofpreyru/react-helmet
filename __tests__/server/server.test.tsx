import type { ReactNode } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import { Helmet } from '../../src';
import { renderContextServer, isArray } from '../../jest/server-utils';

describe('server', () => {
  describe('API', () => {
    it('rewind() provides a fallback object for empty Helmet state', () => {
      const head = renderContextServer(<div />);
      // eslint-disable-next-line jest/no-conditional-in-test
      if (!head) throw Error('Failed');

      expect(head.htmlAttributes).toBeDefined();
      expect(head.htmlAttributes.toString).toBeDefined();
      expect(head.htmlAttributes.toString()).toBe('');
      expect(head.htmlAttributes.toComponent).toBeDefined();
      expect(head.htmlAttributes.toComponent()).toStrictEqual({});

      expect(head.title).toBeDefined();
      expect(head.title.toString).toBeDefined();
      expect(head.title.toString()).toMatchSnapshot();
      expect(head.title.toComponent).toBeDefined();

      const markup = renderToStaticMarkup(
        head.title.toComponent() as unknown as ReactNode,
      );

      expect(markup).toMatchSnapshot();

      expect(head.base).toBeDefined();
      expect(head.base.toString).toBeDefined();
      expect(head.base.toString()).toBe('');
      expect(head.base.toComponent).toBeDefined();

      const baseComponent = head.base.toComponent();

      expect(baseComponent).toStrictEqual(isArray);
      expect(baseComponent).toHaveLength(0);

      expect(head.meta).toBeDefined();
      expect(head.meta.toString).toBeDefined();
      expect(head.meta.toString()).toBe('');
      expect(head.meta.toComponent).toBeDefined();

      const metaComponent = head.meta.toComponent();

      expect(metaComponent).toStrictEqual(isArray);
      expect(metaComponent).toHaveLength(0);

      expect(head.link).toBeDefined();
      expect(head.link.toString).toBeDefined();
      expect(head.link.toString()).toBe('');
      expect(head.link.toComponent).toBeDefined();

      const linkComponent = head.link.toComponent();

      expect(linkComponent).toStrictEqual(isArray);
      expect(linkComponent).toHaveLength(0);

      expect(head.script).toBeDefined();
      expect(head.script.toString).toBeDefined();
      expect(head.script.toString()).toBe('');
      expect(head.script.toComponent).toBeDefined();

      const scriptComponent = head.script.toComponent();

      expect(scriptComponent).toStrictEqual(isArray);
      expect(scriptComponent).toHaveLength(0);

      expect(head.noscript).toBeDefined();
      expect(head.noscript.toString).toBeDefined();
      expect(head.noscript.toString()).toBe('');
      expect(head.noscript.toComponent).toBeDefined();

      const noscriptComponent = head.noscript.toComponent();

      expect(noscriptComponent).toStrictEqual(isArray);
      expect(noscriptComponent).toHaveLength(0);

      expect(head.style).toBeDefined();
      expect(head.style.toString).toBeDefined();
      expect(head.style.toString()).toBe('');
      expect(head.style.toComponent).toBeDefined();

      const styleComponent = head.style.toComponent();

      expect(styleComponent).toStrictEqual(isArray);
      expect(styleComponent).toHaveLength(0);

      expect(head.priority).toBeDefined();
      expect(head.priority.toString).toBeDefined();
      expect(head.priority.toString()).toBe('');
      expect(head.priority.toComponent).toBeDefined();
    });

    it('does not render undefined attribute values', () => {
      const head = renderContextServer(
        <Helmet
          script={[
            {
              async: undefined,
              src: 'foo.js',
            },
          ]}
        />,
      );

      expect(head?.script.toString()).toMatchSnapshot();
    });
  });

  describe('Declarative API', () => {
    it('provides initial values if no state is found', () => {
      const head = renderContextServer(<div />);

      expect(head?.meta).toBeDefined();
      expect(head!.meta.toString).toBeDefined();

      expect(head?.meta.toString()).toBe('');
    });

    // eslint-disable-next-line complexity
    it('rewind() provides a fallback object for empty Helmet state', () => {
      const head = renderContextServer(<div />);

      expect(head?.htmlAttributes).toBeDefined();
      expect(head!.htmlAttributes.toString).toBeDefined();
      expect(head?.htmlAttributes.toString()).toBe('');
      expect(head!.htmlAttributes.toComponent).toBeDefined();
      expect(head?.htmlAttributes.toComponent()).toStrictEqual({});

      expect(head?.title).toBeDefined();
      expect(head!.title.toString).toBeDefined();
      expect(head?.title.toString()).toMatchSnapshot();
      expect(head!.title.toComponent).toBeDefined();

      const markup = renderToStaticMarkup(
        head?.title.toComponent() as unknown as ReactNode,
      );

      expect(markup).toMatchSnapshot();

      expect(head?.base).toBeDefined();
      expect(head!.base.toString).toBeDefined();
      expect(head?.base.toString()).toBe('');
      expect(head!.base.toComponent).toBeDefined();

      const baseComponent = head?.base.toComponent();

      expect(baseComponent).toStrictEqual(isArray);
      expect(baseComponent).toHaveLength(0);

      expect(head?.meta).toBeDefined();
      expect(head!.meta.toString).toBeDefined();
      expect(head?.meta.toString()).toBe('');
      expect(head!.meta.toComponent).toBeDefined();

      const metaComponent = head?.meta.toComponent();

      expect(metaComponent).toStrictEqual(isArray);
      expect(metaComponent).toHaveLength(0);

      expect(head?.link).toBeDefined();
      expect(head!.link.toString).toBeDefined();
      expect(head?.link.toString()).toBe('');
      expect(head!.link.toComponent).toBeDefined();

      const linkComponent = head?.link.toComponent();

      expect(linkComponent).toStrictEqual(isArray);
      expect(linkComponent).toHaveLength(0);

      expect(head?.script).toBeDefined();
      expect(head!.script.toString).toBeDefined();
      expect(head?.script.toString()).toBe('');
      expect(head!.script.toComponent).toBeDefined();

      const scriptComponent = head?.script.toComponent();

      expect(scriptComponent).toStrictEqual(isArray);
      expect(scriptComponent).toHaveLength(0);

      expect(head?.noscript).toBeDefined();
      expect(head!.noscript.toString).toBeDefined();
      expect(head?.noscript.toString()).toBe('');
      expect(head!.noscript.toComponent).toBeDefined();

      const noscriptComponent = head?.noscript.toComponent();

      expect(noscriptComponent).toStrictEqual(isArray);
      expect(noscriptComponent).toHaveLength(0);

      expect(head?.style).toBeDefined();
      expect(head!.style.toString).toBeDefined();
      expect(head?.style.toString()).toBe('');
      expect(head!.style.toComponent).toBeDefined();

      const styleComponent = head?.style.toComponent();

      expect(styleComponent).toStrictEqual(isArray);
      expect(styleComponent).toHaveLength(0);

      expect(head?.priority).toBeDefined();
      expect(head!.priority.toString).toBeDefined();
      expect(head?.priority.toString()).toBe('');
      expect(head!.priority.toComponent).toBeDefined();
    });

    it('does not render undefined attribute values', () => {
      const head = renderContextServer(
        <Helmet>
          <script async={undefined} src="foo.js" />
        </Helmet>,
      );

      expect(head?.script.toString()).toMatchSnapshot();
    });

    it('prioritizes SEO tags when asked to', () => {
      const head = renderContextServer(
        <Helmet prioritizeSeoTags>
          <link
            href="https://www.chipotle.com"
            // eslint-disable-next-line react/no-invalid-html-attribute
            rel="notImportant"
          />
          <link href="https://www.tacobell.com" rel="canonical" />
          <meta content="A very important title" property="og:title" />
          <meta content="Some Description" name="description" />
        </Helmet>,
      );

      expect(head?.priority.toString()).toContain('href="https://www.tacobell.com" rel="canonical"');
      expect(head?.link.toString()).not.toContain('href="https://www.tacobell.com" rel="canonical"');

      expect(head?.priority.toString()).toContain(
        'content="A very important title" property="og:title"',
      );
      expect(head?.priority.toString()).toContain(
        'content="Some Description" name="description"',
      );
      expect(head?.meta.toString()).not.toContain(
        'content="A very important title" property="og:title"',
      );
      expect(head?.meta.toString()).not.toContain(
        'content="Some Description" name="description"',
      );
    });

    it('does not prioritize SEO unless asked to', () => {
      const head = renderContextServer(
        <Helmet>
          <link
            href="https://www.chipotle.com"
            // eslint-disable-next-line react/no-invalid-html-attribute
            rel="notImportant"
          />
          <link href="https://www.tacobell.com" rel="canonical" />
          <meta content="A very important title" property="og:title" />
        </Helmet>,
      );

      expect(head?.priority.toString()).not.toContain(
        'href="https://www.tacobell.com" rel="canonical"',
      );
      expect(head?.link.toString()).toContain('href="https://www.tacobell.com" rel="canonical"');

      expect(head?.priority.toString()).not.toContain(
        'content="A very important title" property="og:title"',
      );
      expect(head?.meta.toString()).toContain(
        'content="A very important title" property="og:title"',
      );
    });
  });
});
