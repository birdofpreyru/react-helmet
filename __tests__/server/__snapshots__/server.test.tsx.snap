// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`server API does not render undefined attribute values 1`] = `"<script data-rh="true" async src="foo.js"></script>"`;

exports[`server API rewind() provides a fallback object for empty Helmet state 1`] = `"<title data-rh="true"></title>"`;

exports[`server API rewind() provides a fallback object for empty Helmet state 2`] = `"<title data-rh="true"></title>"`;

exports[`server Declarative API does not render undefined attribute values 1`] = `"<script data-rh="true" async src="foo.js"></script>"`;

exports[`server Declarative API rewind() provides a fallback object for empty Helmet state 1`] = `"<title data-rh="true"></title>"`;

exports[`server Declarative API rewind() provides a fallback object for empty Helmet state 2`] = `"<title data-rh="true"></title>"`;
